cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -280,22 +280,28 @@ private void addWorkflowNodeState(ProgramId programId, String pid, Map<String, S
       write(key, new RunRecordMeta(record, properties));
     }
   }
+  public void recordProgramInit(ProgramId programId, String pid, long startTs, String twillRunId,
+                                 Map<String, String> runtimeArgs, Map<String, String> systemArgs) {
+    MDSKey.Builder builder = getProgramKeyBuilder(TYPE_RUN_RECORD_STARTED, programId);
+    recordProgramStart(programId, pid, ProgramRunStatus.STARTING, startTs, twillRunId,
+                       runtimeArgs, systemArgs, builder);
+  }
 
   public void recordProgramStart(ProgramId programId, String pid, long startTs, String twillRunId,
                                  Map<String, String> runtimeArgs, Map<String, String> systemArgs) {
     MDSKey.Builder builder = getProgramKeyBuilder(TYPE_RUN_RECORD_STARTED, programId);
-    recordProgramStart(programId, pid, startTs, twillRunId, runtimeArgs, systemArgs, builder);
+    recordProgramStart(programId, pid, ProgramRunStatus.RUNNING, startTs, twillRunId, runtimeArgs, systemArgs, builder);
   }
 
   @VisibleForTesting
   void recordProgramStartOldFormat(ProgramId programId, String pid, long startTs, String twillRunId,
                                    Map<String, String> runtimeArgs, Map<String, String> systemArgs) {
     MDSKey.Builder builder = getVersionLessProgramKeyBuilder(TYPE_RUN_RECORD_STARTED, programId);
-    recordProgramStart(programId, pid, startTs, twillRunId, runtimeArgs, systemArgs, builder);
+    recordProgramStart(programId, pid, ProgramRunStatus.RUNNING, startTs, twillRunId, runtimeArgs, systemArgs, builder);
   }
 
-  private void recordProgramStart(ProgramId programId, String pid, long startTs, String twillRunId,
-                                  Map<String, String> runtimeArgs, Map<String, String> systemArgs,
+  private void recordProgramStart(ProgramId programId, String pid, ProgramRunStatus status, long startTs,
+                                  String twillRunId, Map<String, String> runtimeArgs, Map<String, String> systemArgs,
                                   MDSKey.Builder keyBuilder) {
     String workflowrunId = null;
     if (systemArgs != null && systemArgs.containsKey(ProgramOptionConstants.WORKFLOW_NAME)) {
```

Comment:
So if a MapReduce or Spark program is started by a workflow, the `WorkflowNodeState` will be set to RUNNING. The node status doesn't have a starting state.